Sources/Workspace/Workspace.swift

```diff
@@ -1338,48 +1334,59 @@ extension Workspace {
+            let url: String
             let filter: ProductFilter
         }
 
-        // optimization: preload in parallel
-        let inputDependenciesURLs = inputManifests.compactMap { $0.dependencies.compactMap{ config.mirrors.effectiveURL(forURL: $0.url) } }.flatMap { $0 }
+        // optimization: preload manifest we know about in parallel
+        let inputDependenciesURLs = inputManifests.map { $0.dependencies.map{ config.mirrors.effectiveURL(forURL: $0.url) } }.flatMap { $0 }
         // FIXME: this should not block
-        var loadedManifests = try temp_await { self.loadManifests(forURLs: inputDependenciesURLs, diagnostics: diagnostics, completion: $0) }.reduce(into: [String: Manifest]()) { (partial, manifest) in
-            partial[manifest.url] = manifest
-        }
+        var loadedManifests = try temp_await { self.loadManifests(forURLs: inputDependenciesURLs, diagnostics: diagnostics, completion: $0) }.spm_createDictionary{ ($0.url, $0) }
 
-        let allManifestsWithPossibleDuplicates = try topologicalSort(inputManifests.map({ KeyedPair(($0, ProductFilter.everything), key: NameAndFilter(name: $0.name, filter: .everything)) })) { node in
-            return node.item.0.dependenciesRequired(for: node.item.1).compactMap({ dependency in
+        // continue to load the rest of the manifest for this graph
+        let allManifestsWithPossibleDuplicates = try topologicalSort(inputManifests.map{ KeyedPair($0, key: URLAndFilter(url: $0.url, filter: .everything)) }) { node in
+            return node.item.dependenciesRequired(for: node.key.filter).compactMap{ dependency in
                 let url = config.mirrors.effectiveURL(forURL: dependency.url)
                 // FIXME: this should not block
+                // note: loadManifest emits diagnostics in case it fails
                 let manifest = loadedManifests[url] ?? temp_await { self.loadManifest(forURL: url, diagnostics: diagnostics, completion: $0) }
                 loadedManifests[url] = manifest
-                return manifest.flatMap({ KeyedPair(($0, dependency.productFilter), key: NameAndFilter(name: $0.name, filter: dependency.productFilter)) })
-            })
+                return manifest.flatMap({ KeyedPair($0, key: URLAndFilter(url: $0.url, filter: dependency.productFilter)) })
+            }
         }
 
-        var deduplication: Set<String> = []
-        var allManifests: [KeyedPair<(Manifest, ProductFilter), NameAndFilter>] = []
+        // remove duplicates of the same manifest
+        // FIXME: this drops duplicates with different filters but filter is not in use in this case. we should clean this API up
+        var deduplication: Set<PackageIdentity> = []
+        var allManifests = [(manifest: Manifest, productFilter: ProductFilter)]()
         for node in allManifestsWithPossibleDuplicates {
-            if deduplication.contains(node.item.0.name) {
+            let identity = PackageIdentity(url: node.item.url)
```

Comment:
the second part is to de-dup based on identity instead of name. I actually question the de-duplication especially since its  broken if product filters are used (not a problem now, but will be when they are used)

maybe @aciidb0mb3r has background on why this was added?

Reply:
We could potentially merge the product filters during deduplication. The product filters seem to be used as an input to `precomputeResolution`, so there might indeed be issues there in case the filters make a difference to whether we will perform an actual resolution.

Reply:
I don't really remember. Most likely it just went undetected after we switched to using package identities...

Reply:
In `computePackageURLs()`, we're building up required identities using `PackageReference` which implements equality solely using identity, so whatever we do here should match that.

Reply:
> We could potentially merge the product filters during deduplication. The product filters seem to be used as an input to precomputeResolution, so there might indeed be issues there in case the filters make a difference to whether we will perform an actual resolution.

@neonichu added ptal